### DISCUSSION TITLE
Unable to preserve context in threads since 2.x

### DISCUSSION THREAD

**Discussion by johaven** [pallets/flask#4747] [Q&A]
It is no longer possible to launch a thread in the background which uses the database (since its connection is closed by the context of the application) despite the use `with app.app_context()`

I got this kind of errors:
```python
  File ".venv/lib/python3.8/site-packages/pymysql/connections.py", line 732, in _read_bytes
    data = self._rfile.read(num_bytes)
RuntimeError: reentrant call inside <_io.BufferedReader name=53>
...
pymysql.err.InterfaceError: (0, '')
...
```
Is there another way to preserve the context in some cases?
I use ThreadPoolExecutor patched by gevent for processing tasks after a request.

**Comment by davidism** [comment]
A tread will never have the context of the parent that started it. You should create a new context (and a new connection) _inside_ the thread, not outside it. That was true before 2.2 as well.

```python
def task(app):
    with app.app_context():
        time.sleep(2)
        # do something with the database here
        print(current_app.name)


@bp.route("/start", methods=["GET", "POST"])
def start():
    app = current_app._get_current_object()
    t = Thread(target=task, args=(app,))
    t.start()
    return "started"
```

**Comment by johaven** [comment]
The creation of a new context does not seem really mandatory or blocking, I have also no error message on this subject. It's the closing of the SQL connection that seems random, the threaded tasks run 1 time out of 3, they crash due to the absence of an SQL connection. I'm using your Flask-SqlAlchemy extension (through this simple statement: `db = SQLAlchemy(app, session_options={'expire_on_commit': False})`) and I thought calling `app.app_context()` where the variable is imported from `app = Flask(__name__)` in the init of my code.

However, I found a small workaround that seems to fix these SQL reconnect issues this way:

```python
with app.app_context():
    db.engine.connect() # + to avoid sql connection error
```

Maybe this question should be reported on the Flask-SqlAlchemy repository, sorry.
